packages/eslint-plugin/tests/rules/consistent-type-imports.test.ts

```diff
@@ -1833,5 +1893,447 @@ const b = B;
+      errors: [
+        {
+          messageId: 'inlineTypes',
+          line: 2,
+          column: 9,
+        },
+      ],
+    },
+    {
+      code: `
+        import { A, B } from 'foo';
+        type T = A;
+        B();
+      `,
+      output: `
+        import { type A, B } from 'foo';
+        type T = A;
+        B();
+      `,
+      options: [{ prefer: 'type-imports', fixStyle: 'inline-type-imports' }],
+      errors: [
+        {
+          messageId: 'inlineTypes',
+          line: 2,
+          column: 9,
+        },
+      ],
+    },
+    {
+      code: `
+        import { /* comment */ A, B } from 'foo';
+        type T = A;
+      `,
+      output: `
+        import { /* comment */ type A, B } from 'foo';
+        type T = A;
+      `,
+      options: [{ prefer: 'type-imports', fixStyle: 'inline-type-imports' }],
+      errors: [
+        {
+          messageId: 'inlineTypes',
+          line: 2,
+          column: 9,
+        },
+      ],
+    },
+    {
+      code: `
+        import { B, /* comment */ A } from 'foo';
+        type T = A;
+      `,
+      output: `
+        import { B, /* comment */ type A } from 'foo';
+        type T = A;
+      `,
+      options: [{ prefer: 'type-imports', fixStyle: 'inline-type-imports' }],
+      errors: [
+        {
+          messageId: 'inlineTypes',
+          line: 2,
+          column: 9,
+        },
+      ],
+    },
+    {
+      code: `
+import { A, B, C } from 'foo';
+import type { D } from 'deez';
+
+const foo: A = B();
+let bar: C;
+let baz: D;
+      `,
+      output: `
+import { type A, B, type C } from 'foo';
+import type { D } from 'deez';
+
+const foo: A = B();
+let bar: C;
+let baz: D;
+      `,
+      options: [{ prefer: 'type-imports', fixStyle: 'inline-type-imports' }],
+      errors: [
+        {
+          messageId: 'inlineTypes',
+          line: 2,
+          column: 1,
+        },
+      ],
+    },
+    {
+      code: `
+import { A, B, type C } from 'foo';
+import type { D } from 'deez';
+const foo: A = B();
+let bar: C;
+let baz: D;
+      `,
+      output: `
+import { type A, B, type C } from 'foo';
+import type { D } from 'deez';
+const foo: A = B();
+let bar: C;
+let baz: D;
+      `,
+      options: [{ prefer: 'type-imports', fixStyle: 'inline-type-imports' }],
+      errors: [
+        {
+          messageId: 'inlineTypes',
+          line: 2,
+          column: 1,
+        },
+      ],
+    },
+    {
+      code: `
+import { B } from 'foo';
+import type { A } from 'foo';
+const foo: A = B();
+      `,
+      output: `
+import { B, type A } from 'foo';
+const foo: A = B();
+      `,
+      options: [{ fixStyle: 'inline-type-imports' }],
+      errors: [
+        {
+          messageId: 'inlineTypes',
+          line: 3,
+          column: 1,
+        },
+      ],
+    },
+    {
+      code: `
+import B, { C } from 'foo';
+import type { A } from 'foo';
+
+C();
+const foo: A = B();
+      `,
+      output: `
+import B, { C, type A } from 'foo';
```

Comment:
I think I am seeing what you are suggesting.  I might be all over the place here, but bear with me 😆.  Example at the end that summarizes some thoughts.  

To dedupe `import`/`import type` sources, use [import/no-duplicates](https://github.com/import-js/eslint-plugin-import/blob/376747914b47fbdcf99212b9e9bd4d5e09825385/docs/rules/no-duplicates.md)?  (the one in this repo that extends from this is deprecated it seems)

A few questions that follow.  

1. To dedupe imports, we would need to make sure the eslint rule `import/no-duplicates` could inline the type specifier to an existing non type import source.  Seems like that rule is mainly to merge two sources but not necessarily if one is a value and one is a type import. I'm guessing any additions to this rule would have to assume the "other" dup import was not an `import type` since we wouldn't want the rule to lead to a possible regression and move a type import spec from it's separate type import to a value import.  Which leaves the question, what do we do with the "other" `import type`.  Seems to be in our court?
2. Then, your second suggestion is that on our side, we should prefix a type only specifier in a value import with `type` if a flag is specified?  How does it work that this rule necessarily follows the fixes from `import/no-duplicates`?  Or would a user have to have the right combination of import rules to both inline the import type spec and add the inline `type` prefix? 

e.g.
```js
import { A, B } from 'foo';
import type { Z } from 'foo';
import { R } from 'foo';

type C = B;
A();
let c: Z;
let d: R;

----eslint fixer
import { A, B, R } from 'foo';
import type { Z } from 'foo';

----typescript fixer
import { A, type B, type R } from 'foo';
import type { Z } from 'foo';

----typescript fixer maybe take it this far???  If so, then I think this PR may still apply.
import { A, type B, type R, type Z } from 'foo';
```

To rehash some assumptions I made, `import/no-duplicates` would inline `R` but not `Z` b/c moving Z without the ability to specify the inline `type Z` spec is a regression (assumption that specifying the inline `type` is not in their realm).  Then we would do a pass to add the inline `type R` and inline `type B`.  Lastly, as in the case of the Ember.js repo, we want that `Z` type import to be inlined as well. 

If the eslint import rule would inline `import type { Z }`, I like the simplicity for this repo!  Just add some inline `type` prefixes if necessary.  However, I made a strong/wrong? assumption that doing so on their side might be a possible source of regression.  Let me know what you think and thanks for bearing with me.

Reply:
The thing to think about is that rules should take on one concern at a time.
They should be looking for and solving one problem.
Why? Because otherwise rules blow out and get really complicated - they need many options to help users specify which behaviours they do and don't want.

For `consistent-type-imports` its sole purpose is ensuring that you use type imports (or don't) when an imported name is only used as a type.

Based on the single-purpose mandate, that means that the rule:
1) does not care if you have duplicate imports.
    - this is handled by rules such as `import/no-duplicates`
2) does not care if you use inline vs top-level specifiers.
    - this is, as far as I can tell, not handled by any rules currently. IMO this feature belongs in the import plugin, not here - I have filed https://github.com/import-js/eslint-plugin-import/issues/2469.

---

Now onto the fixer...
In an ideal world the rule would just pick one style to fix to and be done with it, and allow another rule to pick up that import and enforce whether it needs to be deduplicated or inlined or whatever.

However we're not in an ideal world - users want some level of convenience. Hence we're accepting an option to control whether it fixes to a top-level or inline specifier.

What we do *not* want to do, however, is have the fixer be in the business of changing imports that do not violate its single-purpose.

To be clear - if the fix style is set to inline specifier, and there is a top-level specifier - then the rule is not to change it. I.e.

```ts
import type {Foo} from 'T';
import {Bar} from 'T';

type T = Foo | Bar;
```
should fix to
```ts
import type {Foo} from 'T';
import {type Bar} from 'T';

type T = Foo | Bar;
```
and ***NOT***
```ts
import {type Foo, type Bar} from 'T';

type T = Foo | Bar;
```

Reply:
Thanks Brad.  So it sounds like this PR doesn't have much validity and should be closed.  My concern is focused on taking the existing type imports and inlining them with a value import of possible.  That was the desire expressed by folks on the ember.js repo and my company API codebase.  

Based on your thoughts, it seems like `import/no-duplicates` is where this work should go.  However, what was not clear to me is if inlining an existing `import type` to a value import is even something this rule should do. For the practical use case, it necessarily depends on prefixing it with `type` as well.  Otherwise, you are moving an explicit type import to a value import that relies on TS eliding it (which likely happens but doesn't always, thus the explicit type import).  And would rely on pairing this with a rule here.  Is that ok with you?  Just trying to think of the logistics before opening an issue and getting confirmation that I should start work there on `imports/no-duplicates`...

Reply:
To be clear - I think there are 3 rules that should exist:

(1) `@typescript-eslint/consistent-type-imports`

This rule concerns itself with enforcing that your imports that are only used with types are marked as type-only imports.

(2) `import/no-duplicates`

This rule concerns itself with enforcing that you do not have duplicate import declarations for the same module, unless they are differing top-level import kinds.

(3) New rule https://github.com/import-js/eslint-plugin-import/issues/2469

This rule concerns itself with enforcing the style of import kind that you use - whether you always use inline specifiers, or top-level specifiers.

---

To illustrate with an example:

Say we have the following config:
```json5
{
  rules: {
    '@typescript-eslint/consistent-type-imports': 'error',
    'import/no-duplicates': 'error',
    'import/new-rule-issue-2469': ['error', 'prefer-inline'],
  },
}
```

And the following code:
```ts
import type {Foo} from 'T';
import {Bar} from 'T';

type T = Foo | Bar;
```

First pass we would see the following errors:
```ts
import type {Foo} from 'T';
//     ^^^^^^^^^^ Error: prefer inline type specifiers (import/new-rule-issue-2469)
import {Bar} from 'T';
//      ^^^ Error: imports only used as a type must be marked as such (@typescript-eslint/consistent-type-imports)

type T = Foo | Bar;
```
We'd run the fixers for both rules and end up with:
```ts
import {type Foo} from 'T';
import {type Bar} from 'T';

type T = Foo | Bar;
```
This would then cause the following error:
```ts
import {type Foo} from 'T';
import {type Bar} from 'T';
//^^^^^^^^^^^^^^^^^^^^^^^^^ Error: duplicate import from module 'T' (import/no-duplicates)

type T = Foo | Bar;
```
We'd run the fixers and end up with:
```ts
import {type Foo, type Bar} from 'T';

type T = Foo | Bar;
```

Side note to explain how this would really work for a user. They might choose to do things by hand by manually applying the fixer for each rule in their IDE. In this case it'll take 3 manual fixer triggers to reach the code.
If they use autofix-on-save, or if they use the CLI `eslint --fix`, then ESLint will do 10 passes on the file, finding and applying fixes - so all 3 fixes will be applied at once (i.e. takes just 1 user action).

Reply:
Ok.  I like the direction.  😢 a lot of this work is shot but at least you have laid out a more precise integration here.  Will close the PR then and open an issue in `import/no-duplicates`.  Unsure when/if I will be able to get to this.  Also, need to make sure I get buy in.

Reply:
On the other hand, the rub I have with this fixer rule is that it is refactoring my code to an outdated(?) Typescript pattern (separate import types) without the new rule you filed an issue for.  Do you have any thoughts on that @bradzacher?  For example, say the standard recommended pattern was inline type modifier. Wouldn’t we want to do that in the fixer here?

Reply:
> outdated(?)

Nothing outdated about top-level type-specifier for imports.
It's just a matter or personal preference.

> For example, say the standard recommended pattern was inline type modifier. Wouldn’t we want to do that in the fixer here

I think it's definitely a good option to have the fixer that fixes the erroneous named type import to have a type-only specifier.
That's what the issue is about - adding that exact functionality!!

The thing we *don't* want is to merge or change other imports. The only thing the fixer should do is fix the import that's missing a specifier.

To reiterate my previous example:

```ts
import type {Foo} from 'T';
import {Bar} from 'T';

type T = Foo | Bar;
```
should fix to
```ts
import type {Foo} from 'T';
import {type Bar} from 'T';

type T = Foo | Bar;
```
and ***NOT***
```ts
import {type Foo, type Bar} from 'T';

type T = Foo | Bar;
```

I'm more than happy if this rule has an option to fix the flagged import by making it an inline-type-only import instead of making it a separate, top-level type-only import!

Reply:
@bradzacher I was so focused on taking a few codebases from the separate type import refactor back to inline that I didn't realize I could go backwards with `prefer: 'no-type-imports'` and then apply the inline type fixer!  Thanks for the clear direction.  PR is a lot simpler now 👍.  Let me know what you think!

Reply:
Looks like the test suite is still passing.

https://github.com/emberjs/ember.js/pull/20101

Reply:
I believe we have two pending `eslint-plugin-import` PRs that are blocking this work?

https://github.com/import-js/eslint-plugin-import/pull/2475
https://github.com/import-js/eslint-plugin-import/pull/2473